src/schema/v2/filterArtworksConnection.ts

```diff
@@ -575,39 +573,50 @@ const filterArtworksConnectionTypeFactory = (
         )
       }
 
-      const totalPages = computeTotalPages(
-        aggregations.total.value,
-        gravityOptions.size
+      const hasRequestedCollectorSignals = isFieldRequested(
+        "edges.node.collectorSignals",
+        info
       )
 
-      const connection = connectionFromArraySlice(
-        hits,
-        { first, last, after, before },
-        {
-          arrayLength: Math.min(
-            aggregations.total.value,
-            totalPages * gravityOptions.size
-          ),
-          sliceStart: gravityOptions.offset,
-        }
-      )
+      const collectorSignalsRequests = hasRequestedCollectorSignals
+        ? hits.map((artwork) => enrichArtworkWithCollectorSignals(artwork, ctx))
+        : [Promise.resolve()]
 
-      connection.pageInfo.endCursor = pageToCursor(
-        gravityOptions.page + 1,
-        gravityOptions.size
-      )
+      return Promise.all(collectorSignalsRequests).then(() => {
```

Comment:
The way we're implicitly relying on buried mutation here will get confusing in the future for those who didn't write the code, and lead to more complexity. Like, why is there no result being returned from `then`? Thinking... 

Instead, similar to elsewhere, we should stick to inputs and outputs, and keep it simple. If we're firing off a bunch of parallel requests to loaders via this collected array of promise invocations, we should be evaluating the _return results_ of said calls:

```tsx
if (!hasRequestedCollectorSignals) {
  return null
}

const collectorSignalsRequests = hits.map(artwork => getCollectorSignals(artwork, context))

try {
  const collectorSignalsData = await Promise.all(collectorSignalsRequests)
  
  // Do something with the results 
  const hitsWithCollectorSignals = hits.map((artwork, index) => {
    artwork.collectorSignals = collectorSignalsData[index]
    return artwork
  })
  
  const connection = connectionFromArraySlice(hitsWithCollectorSignals, ...)
  
  // etc
}
```

This keeps things more functional and linear, with less hidden operations. (Also, forgive if i'm misunderstanding some of the implementation somewhere!)

Reply:
OK. I tried to split the difference on this one without shuffling thing around too much but I think I got it - the promises do return the artwork and we handle them with `Promise.all(requests).then(artworks => ...` - so the promises we wait for are returning the loaded artworks rather than just saying "I'm still promising here...". LMK if this is ok, hopefully it's otherwise good to merge.